### PR TITLE
Fix issues with setup:install command if scopes are defined in config.php

### PR DESCRIPTION
**Pull request by jalogut**
# Preconditions

* Core Magento, no additional modules
* Tested on 2.2-develop with PHP 7.0 and 2.3-develop with PHP 7.1

# Steps to reproduce

Create `app/etc/config.php` with the following:

- Default store name different than "default". For example "custom"
- One extra store
- run `bin/magento setup:install`

```
return [
    'modules' => [
       /* All core modules enabled */
      ...
    ],
    'scopes' => [
        'websites' => [
            'admin' => [
                'website_id' => '0',
                'code' => 'admin',
                'name' => 'Admin',
                'sort_order' => '0',
                'default_group_id' => '0',
                'is_default' => '0'
            ],
            'base' => [
                'website_id' => '1',
                'code' => 'base',
                'name' => 'Main Website',
                'sort_order' => '0',
                'default_group_id' => '1',
                'is_default' => '1'
            ]
        ],
        'groups' => [
            [
                'group_id' => '0',
                'website_id' => '0',
                'code' => 'default',
                'name' => 'Default',
                'root_category_id' => '0',
                'default_store_id' => '0'
            ],
            [
                'group_id' => '1',
                'website_id' => '1',
                'code' => 'main_website_store',
                'name' => 'Main Website Store',
                'root_category_id' => '2',
                'default_store_id' => '1'
            ]
        ],
        'stores' => [
            'admin' => [
                'store_id' => '0',
                'code' => 'admin',
                'website_id' => '0',
                'group_id' => '0',
                'name' => 'Admin',
                'sort_order' => '0',
                'is_active' => '1'
            ],
            'custom' => [
                'store_id' => '1',
                'code' => 'custom',
                'website_id' => '1',
                'group_id' => '1',
                'name' => 'Default Store View',
                'sort_order' => '0',
                'is_active' => '1'
            ],
            'extra' => [
                'store_id' => '3',
                'code' => 'extra',
                'website_id' => '1',
                'group_id' => '1',
                'name' => 'Extra Store',
                'sort_order' => '1',
                'is_active' => '1'
            ]
        ]
    ]
];
```

# Expected result
Magento being installed and additional store created

# Actual result

## First Error
```
[Progress: 405 / 754]
Module 'Magento_Payment':
[Progress: 406 / 754]
Module 'Magento_CatalogRule':
Installing data... Upgrading data...
[Progress: 407 / 754]
Module 'Magento_CatalogSearch':
Installing data...

  [LogicException]
  There is no such engine:
```

### Reason

- `Magento_CatalogSearch` setup scripts calls `scopeConfig->getValue()` on store scope in one of his internal methods. This requests the value for the default store set in `config.php` which in this example will be `custom`
- Store values are retrieved in the `Fallback` class from DB instead of from the `config.php`:

```
// app/code/Magento/Store/Model/Config/Processor/Fallback.php
public function process(array $data)
{
    if ($this->deploymentConfig->isDbAvailable()) {//read only from db
        $this->storeData = $this->storeResource->readAllStores();
        $this->websiteData = $this->websiteResource->readAllWebsites();
    }
}
```

As we can see, list of stores is here retrieved from the database, where our `custom` store does not exists yet. 

### Solution
- [Code changes in app/code/Magento/Store/Model/Config/Processor/Fallback.php](https://github.com/magento/magento2/pull/16317/files#diff-5dec7e10285c9a96cf5b40b06ece36c0)
- Retrieve lists of stores always from `scopes` object. If I am not wrong, this object contains the list of stores from `config.php` or from the Database if they are not defined in `config.php`

## Second Error
After fixing the previous error, if we try to run `setup:install` again, we get another error at the end.

```
[SUCCESS]: Magento installation complete.
[SUCCESS]: Magento Admin URI: /admin
Processing configurations data from configuration file...

Import failed: Requested store is not found

```

### Reason

- After creating the new `extra` Store, the method `Magento\Store\Model\StoreManager::reinitStores()` is executed. This method calls `$this->scopeConfig->clean()`. Which will first clean the `scopes`.
- When cleaning the scopes, they are also loaded again because one `logging` submethod is calling `scopeConfig->getValue()`.
- When loading the scopes again, the list of stores is different on different objects. 

```
// Magento\Store\App\Config\Type\Scopes
public function get($path = '')
{
    if (null === $this->data) {
        $this->data = new DataObject($this->source->get());
    }
}
```

`$this-data` contains the list of scopes. On `storeRepository` object, this list is still the old one without the new store added.

### Solution

Option 1:
- The right solution will be to clear the `storeRepository` object and `Magento\Config\App\Config\Type\System` before the `Magento\Store\App\Config\Type\Scopes` -> That is not possible however because another race condition is trigger when the `System` is cleared before the `Scopes`

Option 2:
- Do not call `scopeConfig->getValue()` during the `Magento\Store\App\Config\Type\Scopes` to check whether the logs are enabled.

Option 3:
- Execute `app:config:import` in a new thread, where the objects are all clean.
- That is the actual solution of this PR because I did not manage to find a better way after 10 hours working on it. See [Changes on setup/src/Magento/Setup/Console/Command/InstallCommand.php](https://github.com/magento/magento2/pull/16317/files#diff-5dec7e10285c9a96cf5b40b06ece36c0)

## Fixed Issues
1. [#15686](https://github.com/magento/magento2/issues/15686)

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=16317) <br/>All committers have signed the CLA.

**Comment by magento-engcom-team** [comment]
Hi @jalogut. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me {$VERSION} instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://github.com/magento/magento2/wiki/Magento-Contributor-Assistant) documentation

**Review by orlangur — CHANGES_REQUESTED** [review]
Nowadays all changes must be applied to `2.3-develop` first (see https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#rules). Please prepare a new pull request, this one should be `on hold` until the latter is merged.

**Comment by orlangur** [comment]
Closing due to inactivity. Feel free to reach me out anytime later if you wish to continue work on this pull request and it will be reopened.
Thanks for collaboration!

**Comment by jalogut** [comment]
Destination branch has been updated. I guess I can reopen this PR ;)

**Comment by sivaschenko** [comment]
@jalogut I think It's better to [port](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#porting) to 2.3-develop branch as your current code base is 2.2 and is not compatible with 2.3

**Comment by sidolov** [comment]
Hi @jalogut , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Pull request closed without merging**
